tomography_preprocessing/utils/aretomo.py

```diff
@@ -0,0 +1,185 @@
+import os
+from functools import lru_cache
+from pathlib import Path
+from typing import Tuple, Callable, Dict, Any
+
+import numpy as np
+import pandas as pd
+import starfile
+
+from .. import utils
+from ..utils.transformations import S, Rx, Ry, Rz
+
+def read_tlt(file: os.PathLike) -> np.ndarray:
+    """Read an IMOD tlt file into an (n, ) numpy array."""
+    return np.loadtxt(fname=file, dtype=float).reshape(-1)
+
+@lru_cache(maxsize=100)
+
+def get_xf_file(imod_directory: Path, tilt_series_id: str) -> Path:
+    """Get the xf file containing image transforms from an IMOD directory."""
+    return imod_directory / f'{tilt_series_id}.xf'
+
+
+def get_tlt_file(imod_directory: Path, tilt_series_id: str) -> Path:
+    """Get the tlt file containing tilt angles from an IMOD directory."""
+    return imod_directory / f'{tilt_series_id}.tlt'
+
+
+def get_tilt_series_alignment_parameters(
+        imod_directory: Path,
+        tilt_series_id: str
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Get the tilt-series alignment parameters from an IMOD directory.
+
+    Shifts are in pixels and should be applied before rotations.
+    Rotations are ZYZ intrinsic Euler angles which transform the volume
+    """
+    tilt_angles = read_tlt(get_tlt_file(imod_directory,tilt_series_id))  
+    xf = utils.imod.read_xf(get_xf_file(imod_directory,tilt_series_id))
+    shifts_px = utils.imod.get_xf_shifts(xf)
+    in_plane_rotations = utils.imod.get_xf_in_plane_rotations(xf)
+    euler_angles = np.zeros(shape=(len(tilt_angles), 3))
+    euler_angles[:, 1] = tilt_angles
+    euler_angles[:, 2] = in_plane_rotations
+    return shifts_px, euler_angles,
+
+
+def remove_ignored_images(
+    tilt_image_df: pd.DataFrame, 
+    euler_angles: np.ndarray
+) -> pd.DataFrame:
+    """Detect the images removed from the tilt series by AreTomo (due to poor alignment) and remove them from
+    the image stack and star file
+    """
+    star_angles = tilt_image_df['rlnTomoNominalStageTiltAngle']
+    tlt_angles = euler_angles[:,1]
+    idx_min = []
+    for angle in tlt_angles:
+        arr_diff = abs(star_angles - angle)
+        #If difference between nominal tilt and tlt file is less than 0.5, assume same tilt
+        if np.min(arr_diff) < 0.5:	
+            idx_min.append(np.argmin(arr_diff))
```

Comment:
this might be too stringent -> maybe shift to +=20?

Reply:
I thought it might not be stringent enough! This is for comparing the stage tilt with the tilts in the tlt file as I noticed small differences between the mdoc tilt angles and the angles in the tlt file. I hope the stage tilt would not be 20 deg out, unless I'm misinterpreting something here